### DISCUSSION TITLE
Detecting platform using IsOSPlatform() doesn't work in a Docker linux container

### DISCUSSION THREAD

**Discussion by replaysMike** [dotnet/msbuild#8777] [Q&A] [answered]
I'm trying to define the OS platform in my csproj as follows:

```csproj
<PropertyGroup Condition="$([MSBuild]::IsOSPlatform('Windows'))">
    <DefineConstants>OS_WINDOWS</DefineConstants>
</PropertyGroup>
<PropertyGroup Condition="$([MSBuild]::IsOSPlatform('Linux'))">
    <DefineConstants>OS_LINUX</DefineConstants>
</PropertyGroup>
```

which would work fine normally, but when run in a docker linux container it thinks the OS is Windows. Are there any other options?

I've got libraries I'm trying to prevent from being included, but if this is more of a runtime issue then I may need to figure out alternatives.

`RuntimeInformation.IsOSPlatform(OSPlatform.Linux)` works correctly at runtime, but does not for msbuild. In my case I'm using PerformanceCounters when using Windows and not within a container, but I get TypeInitializer issues if it's even referenced at runtime.

```csharp
#if OS_WINDOWS
if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && !EnvironmentUtils.InDocker) {
  var performanceCounter = new PerformanceCounter();
  // no good
}
#endif
```

**Comment by KalleOlaviNiemitalo** [comment, marked as answer]
`Condition="$([MSBuild]::IsOSPlatform('Windows'))"` -- AFAIK this checks the platform on which you run MSBuild.  So if you build on Windows but run on Linux, it won't match.

How do you set the target frameworks; do you have `net6.0-windows` there?

I'm surprised you get a TypeInitializerException.  Which version of the System.Diagnostics.PerformanceCounter package are you using, and what is the inner exception?

**Reply by GangWang01** [reply]
As @KalleOlaviNiemitalo mentioned, `OS_WINDOWS` defined when building on Windows couldn't represent the OS which you run.  `RuntimeInformation.IsOSPlatform(OSPlatform.Linux)` should be used to determine the OS you run.
